internal/transport/controlbuf.go

```diff
@@ -432,6 +431,14 @@ func (c *controlBuffer) finish() {
 			hdr.onOrphaned(ErrConnClosing)
 		}
 	}
+	// In case throttle() is currently in flight, it needs to be unblocked.
+	// Otherwise, the transport may not close, since the transport is closed by
+	// the reader encountering the connection error.
```

Comment:
Also add this?

This won't race with any other goroutine, because the caller (the `loopyWriter()` goroutine) is the only goroutine that ever sets the channel

Reply:
Races are actually avoided by taking the lock and setting `c.err`, and `get`, which sets the channel, checking the error.  This is because `finish` may also be called by `t.Close` while `get` is in flight.

Reply:
Oh, `t.Close()` also calls `finish()`? Why would it need to?

Reply:
I don't think it needs to now.  I can remove it, but I don't think it simplifies things, really.

Actually, I think we could change things a little bit to improve them slightly.

1. Call `finish` from a `defer` in `run`.
2. Change `trfChan` to a `chan struct{}` instead of `atomic.Value` and access it with `atomic.___Pointer`
3. `finish` could close `trfChan` directly without even the atomic load, since we know `finish` is only called by `run`.

Reply:
And the "race" I'm talking about here is not read/write race.
It's that if `finish()` unblocks this channel, but another goroutine sets it to block again, before the `reader()` has a chance to unblock, leading to another deadlock.

Reply:
There is no read/write race due to the use of the atomic.  I think we were talking about the same thing.  And if `finish` is called by `Close`, it can happen while `get` is in flight, which, if `get` didn't check `c.err`, could encounter this race.  Anyway, let me know what you want me to do here.

Reply:
The change in this PR is OK.

We should cleanup the transport code later (when? :man_shrugging:).
- the channel (`controlbuf`) is closed by the reader (`loopyWriter()`) instead of the writer (`reader()`)     (hahahahaha)
  - which is, counterintuitive
- the multiple caller of `t.Close()` is probably part of the cause that RPCs tend to fail with `transport is closing`
  - if the transport is only closed by the `reader()`, and the other places just tells `reader()` to close, it might make things cleaner

Reply:
#4459